### PR TITLE
Fix: avoid using `process.exit()` when possible

### PR DESCRIPTION
**Pull request by BYK**
**Summary**

Refs #3524. We are using `process.exit()` pretty liberally although it is [strongly advised not to use it](https://nodejs.org/api/process.html#process_process_exit_code) since it may cause the `stdout` to get corrupted/terminated before a full flush. This patch changes all possible `process.exit(code)` calls with `process.exitCode = code` statements.

We'd also ideally enable [no-process-exit](http://eslint.org/docs/rules/no-process-exit) rule in ESLint but it requires an upgrade to ESLint v4 which should be handled separately.

**Test plan**

Expect tests to pass and actually finish (not run indefinitely due to `yarn` not exiting at all). Also, the script referenced in #3524 to output something like the following:
```
DATA 18
DATA 8192
DATA 8192
DATA 5392
DATA 15
EXIT 0
```

### CODE REVIEW

File: src/cli/index.js [hunk 1]
```diff
@@ -142,7 +142,7 @@ if (outputWrapper) {
 if (command.noArguments && commander.args.length) {
   reporter.error(reporter.lang('noArguments'));
   reporter.info(command.getDocsInfo);
-  process.exit(1);
+  process.exitCode = 1;
```
Comment (arcanis) [inline src/cli/index.js]: You need to shortcut the rest of the code to have the same behavior - `process.exit` doesn't return, but setting `process.exitCode` will continue executing the rest of the code.

File: src/util/signal-handler.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ import {forwardSignalToSpawnedProcesses} from './child.js';
 
 function forwardSignalAndExit(signal: string) {
   forwardSignalToSpawnedProcesses(signal);
-  process.exit(1);
+  throw new Error('Received SIGTERM, terminating.');
```
Comment (arcanis) [inline src/util/signal-handler.js]: Will that print a stacktrace on the console? That won't be helpful for the user.

Comment (BYK) [inline src/util/signal-handler.js]: I thought it would be. Why do you think it wouldn't?

File: src/cli/index.js (hunk 1 above)
Comment (BYK) [inline src/cli/index.js]: Not shortcutting the whole point with this change. We should not need the shortcut behavior and turns out we really don't since all tests are passing. It was only necessary at one point so I left it that way.

File: src/util/signal-handler.js (hunk 2 above)
Comment (arcanis) [inline src/util/signal-handler.js]: `SIGTERM` is the default signal sent by `kill`. It doesn't make much sense to print a stacktrace (which heavily imply that something went wrong in the program implementation) if the termination came from the outside, don't you think?

Comment (BYK) [inline src/util/signal-handler.js]: Alright, makes sense. I'll find an alternative way for this one.

File: src/cli/index.js (hunk 1 above)
Comment (arcanis) [inline src/cli/index.js]: I don't think that's a good idea.

- it kinda violates the principle of the least surprise, since we would no longer have the guarantee that any part of this file would only be executed if no other command matched. It's only a matter of time before we make some mistake that will execute two commands in some edge case.

- Case in point, your current code has a bug that will make `yarn -v` print the version number before running `yarn install -v` (and `yarn -v -h` will print the version number then call `yarn help -v`).

- Even without that, not shortcutting also means that we will be running useless tasks (such as instanciating a reporter etc) even when not needed. That's an issue because those commands must be fast (some tools such as Nuclide might need to call them often, and if their results are late it will slow down the all process).

I'm not against using `exitCode`, you have a point that shutting down the program before the I/O has finished is bad, but I strongly think we then need to refactor the CLI code to make sure we never have to check that our code won't get called twice - might only be a matter of adding `else` clauses 😃

File: src/cli/index.js [hunk 3]
```diff
@@ -20,347 +20,352 @@ const net = require('net');
...
-// if -v is the first command, then always exit after returning the version
-if (commandName === '-v') {
-  console.log(version.trim());
-  process.exitCode = 0;
-}
+export default function main() {
+  loudRejection();
+  handleSignals();
+
+  const startArgs = process.argv.slice(0, 2);
```
Comment (arcanis) [inline src/cli/index.js]: What do you think about modularizing and putting this in the `main` parameters? I can see this function as an interesting entry point to run commands in our testcases - we currently need to instanciate each class and pass them the right properties, which might be tedious and error-prone, so it could be nice to just be able to use `run([ 'install', '--check-files' ])` instead and use the actual CLI pipeline.

Comment (arcanis) [inline src/cli/index.js]: Actually my comment was more about [Line 31](https://github.com/yarnpkg/yarn/pull/3955/files#diff-867becf4a9c2c6c6d4e7c1278750724eR31), but it's also applicable here.

File: src/reporters/console/console-reporter.js [hunk 4]
```diff
@@ -196,7 +196,7 @@ export default class ConsoleReporter extends BaseReporter {
         (err, answer) => {
           if (err) {
             if (err.message === 'canceled') {
-              process.exit(1);
```
Comment (arcanis) [inline src/reporters/console/console-reporter.js]: Shouldn't we exit immediately here? Or reject?

File: scripts/build-webpack.js [hunk 5]
```diff
@@ -35,6 +35,8 @@ const compiler = webpack({
   output: {
     filename: `yarn-${version}.js`,
     path: path.join(basedir, 'artifacts'),
+    library: 'yarn-cli',
+    libraryTarget: 'commonjs2',
```
Comment (arcanis) [inline scripts/build-webpack.js]: Is this change required? What breaks without it?

Comment (BYK) [inline scripts/build-webpack.js]: Yes, this is the solution that fixes the single-file bundles. Without this, Webpack exports an empty object, causing `yarn-bundled-entry-point.js` to fail since `main` function can't be imported.

File: src/reporters/console/console-reporter.js (hunk 4 above)
Comment (BYK) [inline src/reporters/console/console-reporter.js]: Yeah I think you're right. I thought about the same thing and don't know why I didn't merge the `reject` part. Update coming.

**Comment by skevy** [comment]
I noticed when pulling down master and running `build-dist` that the webpack builds were no longer working.

I traced it back to this PR. You can see in CircleCI the output of the last working build here:

https://circleci.com/gh/yarnpkg/yarn/4533

Notice how `$ ./artifacts/yarn-`./dist/bin/yarn --version`.js --version` outputs a version number.

When this commit runs in CI, however (`https://circleci.com/gh/yarnpkg/yarn/4536`), there is no version number output, and I confirm that the webpack build does not work at all.

**Comment by Daniel15** [comment]
Hmm, interesting, I wonder why the nightly end-to-end tests didn't fail. They use the nightly Debian packages, which contain the Webpack build inside them. https://build.dan.cx/job/yarn-e2e/label=docker,os=ubuntu-16.04/269/console

**Comment by skevy** [comment]
I'm not sure why those don't fail (without looking into it more) -- but the reason CircleCI doesn't fail is because `yarn --version` still exists with code 0, even though it doesn't do anything.

**Comment by Daniel15** [comment]
Right. Maybe the build should assert that it actually returns the correct version number.

**Comment by skevy** [comment]
👍

**Comment by BYK** [comment]
I'll work on this soon.

**Comment by BYK** [comment]
Filed #4057, actively working on it.
